File: src/Shared/Students.Models/EducationProgram.cs
using System.Text.Json.Serialization;
using Students.Models.ReferenceModels;

namespace Students.Models;

/// <summary>
/// Образовательная программа.
/// </summary>
public class EducationProgram
{
  /// <summary>
  /// Id программы.
  /// </summary>
  public Guid Id { get; set; }

  /// <summary>
  /// Наименование программы.
  /// </summary>
  public string? Name { get; set; }

  /// <summary>
  /// Стоимость обучения.
  /// </summary>
  public required double Cost { get; set; }

  /// <summary>
  /// Количество часов.
  /// </summary>
  public required int HoursCount { get; set; }

  /// <summary>
  /// Id Формы обучения.
  /// </summary>
  public required Guid EducationFormId { get; set; }

  /// <summary>
  /// Вид документа повышения квалификации.
  /// </summary>
  public required Guid KindDocumentRiseQualificationId { get; set; }

  /// <summary>
  /// Вид программы
  /// </summary>
  public required Guid KindEducationProgramId { get; set; }

  /// <summary>
  /// Модульная программа.
  /// </summary>
  public required bool IsModularProgram { get; set; }

  /// <summary>
  /// Id ВЭД программы.
  /// </summary>
  public Guid? FEAProgramId { get; set; }

  /// <summary>
  /// Id источника финансирования.
  /// </summary>
  public required Guid FinancingTypeId { get; set; }

  /// <summary>
  /// Обязательно наличие ВО (Вот это наследие от прежних разрабов - похоже не нужно).
  /// </summary>
  public required bool IsCollegeProgram { get; set; }

  /// <summary>
  /// Архивная программа.
  /// </summary>
  public required bool IsArchive { get; set; }

  /// <summary>
  /// Сетевая форма.
  /// </summary>
  public required bool IsNetworkProgram { get; set; }

  /// <summary>
  /// Применение ДОТ.
  /// </summary>
  public required bool IsDOTProgram { get; set; }

  /// <summary>
  /// Применение ДОТ полностью.
  /// </summary>
  public required bool IsFullDOTProgram { get; set; }

  /// <summary>
  /// Наименование квалификации, профессии, специальности
  /// </summary>
  public required string QualificationName { get; set; }

  /// <summary>
  /// Форма обучения.
  /// </summary>
  [JsonIgnore]
  public virtual EducationForm? EducationForm { get; set; }

  /// <summary>
  /// Вид документа повышения квалификации.
  /// </summary>
  [JsonIgnore]
  public virtual KindDocumentRiseQualification? KindDocumentRiseQualification { get; set; }

  /// <summary>
  /// Вид программы повышения квалификации
  /// </summary>
  [JsonIgnore]
  public virtual KindEducationProgram? KindEducationProgram { get; set; }

  /// <summary>
  /// ВЭД программы.
  /// </summary>
  [JsonIgnore]
  public virtual FEAProgram? FEAProgram { get; set; }

  /// <summary>
  /// Источник финансирования.
  /// </summary>
  [JsonIgnore]
  public virtual FinancingType? FinancingType { get; set; }

  /// <summary>
  /// Группы обучения.
  /// </summary>
  [JsonIgnore]
  public virtual ICollection<Group> Groups { get; set; } = new List<Group>();

  /// <summary>
  /// Заявки.
  /// </summary>
  [JsonIgnore]
  public virtual ICollection<Request> Requests { get; set; } = new List<Request>();
}

File: src/Client/src/storage/form-model/education-program.ts
import { ArchiveCheckbox } from '../../components/shared/control/archive-check-box';
import { Cost } from '../../components/shared/control/cost';
import { HoursCount } from '../../components/shared/control/hours-count';
import { EducationFormSelect } from '../../components/shared/control/selects/education-form-select';
import { FEAProgramSelect } from '../../components/shared/control/selects/fea-program-select';
import { FinancingTypeSelect } from '../../components/shared/control/selects/financing-type-select';
import { KindDocumentRiseQualificationSelect } from '../../components/shared/control/selects/kind-document-rise-qualification-select';
import { KindEducationProgramSelect } from '../../components/shared/control/selects/kind-education-program-select';
import { StringControl } from '../../components/shared/control/string-control';
import { YesNoControl } from '../../components/shared/control/yes-no-control';
import { DtoKeys } from '../service/types';
import { FormModel } from './types';

export const educationProgramFormModel: FormModel = {
  name: {
    name: 'Программа обучения',
    type: StringControl,
    formParams: {
      key: DtoKeys.NAME,
      rules: [
        {
          required: true,
        },
      ],
    },
  },
  hoursCount: {
    name: 'Кол-во часов',
    type: HoursCount,
    formParams: {
      key: DtoKeys.HOURS_COUNT,
      rules: [
        {
          required: true,
        },
      ],
    },
  },
  educationFormId: {
    name: 'Форма образования',
    type: EducationFormSelect,
    formParams: {
      key: DtoKeys.EDUCATION_FORM_ID,
      rules: [
        {
          required: true,
        },
      ],
    },
  },
  kindEducationProgramId: {
    name: 'Вид программы',
    type: KindEducationProgramSelect,
    formParams: {
      key: DtoKeys.KIND_EDUCATION_PROGRAM_ID,
      rules: [
        {
          required: true,
        },
      ],
    },
  },
  cost: {
    name: 'Стоимость',
    type: Cost,
    formParams: {
      key: DtoKeys.COST,
      rules: [
        {
          required: true,
        },
      ],
    },
  },
  financingTypeId: {
    name: 'Источник финансирования',
    type: FinancingTypeSelect,
    formParams: {
      key: DtoKeys.FINANCING_TYPE_ID,
      rules: [
        {
          required: true,
        },
      ],
    },
  },
  isModularProgram: {
    name: 'Модульная программа',
    type: YesNoControl,
    formParams: {
      key: DtoKeys.IS_MODULAR_PROGRAM,
      rules: [
        {
          required: true,
        },
      ],
    },
  },
  feaProgramId: {
    name: 'ВЭД программы',
    type: FEAProgramSelect,
    formParams: {
      key: DtoKeys.FEA_PROGRAM_ID,
      rules: [
        {
          required: true,
        },
      ],
    },
  },
  isNetworkProgram: {
    name: 'Сетевая форма',
    type: YesNoControl,
    formParams: {
      key: DtoKeys.IS_NETWORK_PROGRAM,
      rules: [
        {
          required: true,
        },
      ],
    },
  },
  isDOTProgram: {
    name: 'Применение ДОТ',
    type: YesNoControl,
    formParams: {
      key: DtoKeys.IS_DOT_PROGRAM,
      rules: [
        {
          required: true,
        },
      ],
    },
  },
  isFullDOTProgram: {
    name: 'Применение ДОТ полностью',
    type: YesNoControl,
    formParams: {
      key: DtoKeys.IS_FULL_DOT_PROGRAM,
      rules: [
        {
          required: true,
        },
      ],
    },
  },
  isCollegeProgram: {
    name: 'Обязательно наличие ВО',
    type: YesNoControl,
    formParams: {
      key: DtoKeys.IS_COLLEGE_PROGRAM,
      rules: [
        {
          required: true,
        },
      ],
    },
  },
  qualificationName: {
    name: 'Наименование квалификации',
    type: StringControl,
    formParams: {
      key: DtoKeys.QUALIFICATION_NAME,
      rules: [
        {
          required: true,
        },
      ],
    },
  },
  kindDocumentRiseQualificationId: {
    name: 'Вид документа',
    type: KindDocumentRiseQualificationSelect,
    formParams: {
      key: DtoKeys.KIND_DOCUMENT_RISE_QUALIFICATION_ID,
      rules: [
        {
          required: true,
        },
      ],
    },
  },
  inArchive: {
    name: 'В архиве',
    type: ArchiveCheckbox,
    formParams: {
      key: DtoKeys.IS_ARCHIVE,
      rules: [
        {
          required: true,
        },
      ],
    },
  },
};
